File: common/src/main/java/juuxel/adorn/client/resources/ColorPalette.java
package juuxel.adorn.client.resources;

import juuxel.adorn.AdornCommon;
import net.minecraft.resources.Identifier;

import java.util.Map;

public final class ColorPalette {
    private static final Identifier DEFAULT_COLOR_ID = AdornCommon.id("default");
    private final Map<Identifier, ColorManager.ColorPair> map;

    public ColorPalette(Map<Identifier, ColorManager.ColorPair> map) {
        this.map = Map.copyOf(map);
    }

    public ColorManager.ColorPair get(Identifier key) {
        var pair = map.get(key);
        if (pair == null) {
            pair = map.get(DEFAULT_COLOR_ID);
            if (pair == null) throw new IllegalStateException("Couldn't read default value from palette map");
        }
        return pair;
    }
}

File: common/src/main/kotlin/juuxel/adorn/client/resources/ColorPalette.kt
package juuxel.adorn.client.resources

import com.google.common.collect.ImmutableMap
import juuxel.adorn.AdornCommon
import net.minecraft.util.Identifier

class ColorPalette(map: Map<Identifier, ColorManager.ColorPair>) {
    private val map = ImmutableMap.copyOf(map)

    operator fun get(key: Identifier): ColorManager.ColorPair =
        map.getOrElse(key) {
            map[DEFAULT_COLOR_ID] ?: error("Couldn't read default value from palette map")
        }

    companion object {
        private val DEFAULT_COLOR_ID = AdornCommon.id("default")
    }
}
